### PR TITLE
Initial commit of MLIR function builder

### PR DESCRIPTION
**Pull request by paul-tqh-nguyen**
This is the initial implementation of the IR builder. 

Here's how we can implement triangle counting:
```
    # Grab functions used by triangle count
    transpose_function = Transpose()
    matrix_select_triu_function = MatrixSelect("TRIU")
    matrix_select_tril_function = MatrixSelect("TRIL")
    matrix_reduce_function = MatrixReduceToScalar()
    mxm_plus_pair_function = MatrixMultiply("plus_pair", mask=True)

    # Inputs to triangle count (includes sparse tensors to store intermediate values)
    A_var = MLIRBuilderVar("A", "!llvm.ptr<i8>")
    U_var = MLIRBuilderVar("U", "!llvm.ptr<i8>")
    L_var = MLIRBuilderVar("L", "!llvm.ptr<i8>")
    UT_var = MLIRBuilderVar("UT", "!llvm.ptr<i8>")
    C_var = MLIRBuilderVar("C", "!llvm.ptr<i8>")

    # Initialize the IR builder with the input variables and return type 
    # (zero or multiple return types are WIP)
    ir_builder = MLIRFunctionBuilder("triangle_count", A_var, L_var, U_var, UT_var, C_var, return_type="f64")

    # Call the functions / implement triangle count
    ir_builder.add_function_call(matrix_select_triu_function, U_var, A_var)
    ir_builder.add_function_call(matrix_select_tril_function, L_var, A_var)
    ir_builder.add_function_call(transpose_function, UT_var, U_var)
    ir_builder.add_function_call(mxm_plus_pair_function, C_var, L_var, UT_var, L_var)
    reduce_result = ir_builder.add_function_call(matrix_reduce_function, C_var)

    # Return the answer
    ir_builder.add_return_statement(reduce_result)

    # Tell the IR builder to compile for us a Python callable
    triangle_count_internal = ir_builder.compile(engine=engine)

    # A wrapper function (needed to allocate sparse tensors for intermediate values)
    def triangle_count(A: MLIRSparseTensor) -> int:
        U = create_empty_matrix(*A.shape, nnz=A.pointers[1][-1])
        L = create_empty_matrix(*A.shape, nnz=A.pointers[1][-1])
        UT = create_empty_matrix(*U.shape, nnz=U.pointers[1][-1])
        C = create_empty_matrix(*A.shape, nnz=L.pointers[1][-1])
        answer = triangle_count_internal(A, U, L, UT, C)
        answer = int(answer)
        return answer
```

There are tests in `mlir-graphblas/mlir_graphblas/tests/test_mlir_builder.py` that demonstrate how to use the IR builder. 

This PR also refactors `engine.py` (intended to be a no-op).

### CODE REVIEW

File: mlir_graphblas/tests/test_mlir_builder.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+    input_tensor = MLIRSparseTensor(indices, values, sizes, sparsity)
+    output_tensor = create_empty_matrix(
+        *input_tensor.shape, nnz=input_tensor.pointers[1][-1]
+    )
+
+    dense_input_tensor = np.zeros([8, 8], dtype=np.float64)
+    dense_input_tensor[1, 2] = 1.2
+    dense_input_tensor[4, 3] = 4.3
+    assert np.isclose(dense_input_tensor, engine.densify(input_tensor)).all()
```
Comment (jim22k) [inline mlir_graphblas/tests/test_mlir_builder.py:107]: This is redundant with the statement two lines down

**Review by jim22k — CHANGES_REQUESTED** [review]
Overall looks really good.

Eventually, we may want to incorporate the `MLIRFunc` object from `metagraph_mlir/compiler.py`. But that can wait for a future refactor.

File: mlir_graphblas/tests/test_mlir_builder.py (hunk 1 above)
Comment (paul-tqh-nguyen) [reply mlir_graphblas/tests/test_mlir_builder.py:107]: I can remove this line if it's unnecessary. My initial motivation was to make sure that the MLIR function didn't accidentally change the contents of `input_tensor`. That might be excessive.

Comment (jim22k) [reply mlir_graphblas/tests/test_mlir_builder.py:107]: Ah, makes sense. I didn't understand why the same test was being run twice, but checking for no mutations is valid. Just add a comment on both lines explaining what is being tested.

File: mlir_graphblas/mlir_builder.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+from .engine import parse_mlir_string
+
+from typing import Dict, List, Tuple
+
+#############
+# IR Builer #
+#############
+
+
+class MLIRBuilderVar:
```
[the remark this answers is not in the document]
Comment (paul-tqh-nguyen) [reply mlir_graphblas/mlir_builder.py]: https://github.com/metagraph-dev/mlir-graphblas/pull/44/commits/9b55b48e23697775fab554e21c157ec7012de6dd#diff-e75c7d649301c8f462d9eb17ef79ed5bdbb5883a7a63c1764281d2d35d077038R22 implements this renaming.

File: mlir_graphblas/mlir_builder.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            statements=joined_statements,
+        )
+
+    def add_return_statement(self, returned_value: MLIRBuilderVar) -> None:
+        returned_value_string = (
+            returned_value.var_string
+            if isinstance(returned_value, MLIRBuilderVar)
+            else str(returned_value)
+        )
+        statement = f"return {returned_value_string} : {self.return_type}"
```
[the remark this answers is not in the document]
Comment (paul-tqh-nguyen) [reply mlir_graphblas/mlir_builder.py:95]: https://github.com/metagraph-dev/mlir-graphblas/pull/44/commits/9b55b48e23697775fab554e21c157ec7012de6dd#diff-e75c7d649301c8f462d9eb17ef79ed5bdbb5883a7a63c1764281d2d35d077038R86 adds this sanity check.

File: mlir_graphblas/tests/test_mlir_builder.py (hunk 1 above)
Comment (paul-tqh-nguyen) [reply mlir_graphblas/tests/test_mlir_builder.py:108]: https://github.com/metagraph-dev/mlir-graphblas/pull/44/commits/9b55b48e23697775fab554e21c157ec7012de6dd#diff-8a0c1e6dbb5e24631939b88ee63968ef6e90bdad26c8539d2e0af6c193af2864R112 adds the comment.

File: mlir_graphblas/mlir_builder.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        if function in self.needed_function_table:
+            function_mlir_text, input_types, return_type = self.needed_function_table[
+                function
+            ]
+        else:
+            function_mlir_text = function.get_mlir(make_private=True)
+            (function_ast,) = parse_mlir_string(function_mlir_text).body
+            input_types = [arg.type.dump() for arg in function_ast.args]
+            return_type = function_ast.result_types.dump()
+            self.needed_function_table[function] = (
```
[the remark this answers is not in the document]
Comment (paul-tqh-nguyen) [reply mlir_graphblas/mlir_builder.py]: https://github.com/metagraph-dev/mlir-graphblas/pull/44/commits/9b55b48e23697775fab554e21c157ec7012de6dd#diff-8a0c1e6dbb5e24631939b88ee63968ef6e90bdad26c8539d2e0af6c193af2864R118 implements this suggestion. There's now a test for this situation in `test_ir_builder_double_transpose`.

**Pull request merged**